[RDY][CR] Refactor morale (using Observer), part 3

codemime opened the pull request:
Continues #15353. Intentions: decouple morale stuff and keep most of it hidden behind a narrow interface; mitigate size of god-like `player` (even a bit); make pieces of code a bit cleaner.

Fixes #15408

Trying to implement a back pointer to the `player` I've encountered an obstacle. Since the objects are copied, it's hard to sustain back pointer's correctness and preserve default copy/move constructors at the same time.

I see four possible solutions here (from the worst):
1. Non-default copy/move constructors/operations.
   Drawbacks: troublesome and error-prone (something can be forgotten).
   Benefits: explicit.
2. Ancestor class (kinda mixin) that handles back pointer:

``` c++
class morale_subject
{
    public:
        morale_subject() : morale( ( player *)this ) {};
        morale_subject( const morale_subject &another ) : morale( another.morale ) {
             morale.p = ( parent * )this;
        }
        owner &operator = ( const morale_subject &another ) {
             morale = another.morale;
             morale.p = ( parent * )this;
             return *this;
        }

        child morale;
};

class player: public Character, public morale_subject, ...
{
    ...
```

Drawbacks: hard cast, increases complexity, requires multiple inheritance
Benefits: launch and probably forget
3. Hardcoded `g->u`
Drawbacks: inflexibility and intolerance to further changes
Benefits: simplicity
4. Passing the ~~pointer~~ ~~reference (chosen)~~
Drawbacks: not so elegant
Benefits: absence of other drawbacks
5. Passing none (chosen).

~~Maybe there's a much better solution which I don't see yet (some kind of sophisticated smart pointer or whatever)?~~ There is.

Also, during simplification I found a dead piece of code inside `apply_persistent_morale()`. Should I remove it or it meant something?

``` c++
for( auto &elem : worn ) {
    if( elem.has_flag( basic_flag ) || elem.has_flag( bonus_flag ) ) {
        covered |= elem.get_covered_body_parts(); <--the cause
    }
    if( elem.has_flag( bonus_flag ) ) {
        bonus+=2;
    } else if( elem.has_flag( basic_flag ) ) {
        if( ( covered & elem.get_covered_body_parts() ).none() ) {  <-- always false
            bonus += 1;  <-- dead increment
        }
    }
}
```

I've ~~temporarily commented out~~ removed the assignment in the simplified version.

BevapDin commented on src/savegame_json.cpp:
```diff
@@ -2004,3 +2001,15 @@ void morale_point::serialize( JsonOut &json ) const
+
+void player_morale::load( JsonObject &jsin )
+{
+    points.clear();
```
There is no need to clear this explicitly, `JsonObject::clear` already does it.

BevapDin commented on src/morale.cpp:
```diff
@@ -137,10 +218,379 @@ void morale_point::proceed( int ticks )
+    level_is_valid = false;
+}
+
+int player_morale::get_hoarder_penalty( const player *p ) const
```
You should use references (const if possible) instead of pointers. Your code seems to assume the pointer is always non-null (no checks for null). Using references in that situation makes that clear to the caller of those functions.

codemime replied:
Agree.

kevingranade commented:
Why are you trying to store a back pointer to the player in the first place? Ideally the player_morale class won't know anything about the player.  Also as you've noticed it's hard to keep pointers consistent over time, it's much better to just avoid them.

codemime commented:
Generally I agree, but it has to know about the `player` in order to calculate morale bonuses and penalties. IMHO that shouldn't be `player`'s responsibility, it already has too many. Do you have an idea of how to decouple them properly? I think of another independent class with static methods (kinda flyweight) that gets the `player` (and possibly NPCs in the future?) and repopulates `player.morale`. Then the `player_morale` class can become a plain container.

Coolthulhu commented:
> Do you have an idea of how to decouple them properly?

Have player methods call its methods with a pointer to self, like this;
```
void player::update_some_stuff()
{
    morale_container.update( *this );
}
```

codemime commented:
@Coolthulhu That's exactly what I've done.

Coolthulhu commented:
I don't see the problem then.

The idea with keeping a back pointer sounds just too error prone to me. I don't see a good way of doing it without changing the entire class based approach we have to a global id database kind of a thing.

> I think of another independent class with static methods (kinda flyweight) that gets the player (and possibly NPCs in the future?) and repopulates player.morale. Then the player_morale class can become a plain container.

I see it de-bloating player.h (and indirectly everything including it), but I don't get the flyweight part. You want the object to keep a pointer to player, to be a wrapper stored in player or what?

What's the motivation? Just avoiding passing the player reference every time?

codemime commented:
> You want the object to keep a pointer to player, to be a wrapper stored in player or what?

Neither. Just an independent class with static methods, to call it like this in the (for example) `do_turn()` method:  `morale_bonuses.apply_to( u );`

> What's the motivation? Just avoiding passing the player reference every time?

Well, yes. It was a possible answer to Kevin's suggestion:

> Ideally the player_morale class won't know anything about the player

codemime commented:
> The idea with keeping a back pointer sounds just too error prone to me

Just so. Although it wouldn't be a problem with non-copyable classes.

kevingranade commented:
What you generally want to do is have the player class "tell" the morale class about updates it cares about.  I'll point out some examples in the code.
An acid test for decoupling is "can I instantiate this class by itself"?  In this case the answer is no, you can't do anything with a player_morale instance without instantiating a player.
*hint* a practical version of the test is to write a unit test *hint*

kevingranade commented on src/morale.cpp:
```diff
@@ -137,10 +218,379 @@ void morale_point::proceed( int ticks )
+    }
+}
+
+morale_mult player_morale::get_traits_mult( const player &p ) const
```
Instead of passing a player here, the player_morale class can cache whether it has these bonuses, and the player can tell the player_morale class whenever it gains or loses a trait (all of them, not just the ones the player_morale class cares about).
This is an example of the [observer pattern](http://gameprogrammingpatterns.com/observer.html), except with only one listener, eventually we might make a full-blown observer class to help out with this kind of thing.

kevingranade commented on src/morale.cpp:
```diff
@@ -137,10 +218,379 @@ void morale_point::proceed( int ticks )
+    return ret;
+}
+
+morale_mult player_morale::get_effects_mult( const player &p ) const
```
Same here, in player::add_effect() you can do ```player_morale.gained_effect( effect_name )```.

kevingranade commented:
> Just so. Although it wouldn't be a problem with non-copyable classes.

Fragility aside, back pointers are a bad code smell, when composing classes, the lower-level classes shouldn't be querying the upper-level classes about their state, the upper-level classes should be telling the lower level classes about it using an API that models what the class wants to know.

codemime commented:
You're right. Observer the cleanest solution here, I'll implement it without subscription. Thank you.

codemime commented:
However there's another pitfall. Since instances serialize/deserialize themselves independently and don't know a thing about their owners, each of them should also serialize it's cached state (which is sloppy - duplication of stored data), or the owner should "reapply" them in the correct order right after deserialization to reconstruct the whole state (that's what I would call fragility). It seems to me that in case of chosen architecture observer-based approach complicates things a bit more than it simplifies them.

kevingranade commented:
Does the order in which traits/effects are acquired affect the outcome? If
so I'd say it's the formula that's fragile, the player should just be able
to replay all traits and effects into a morale instance and get the same
outcome.

codemime commented:
> Does the order in which traits/effects are acquired affect the outcome?

It's not necessary so in case of morale, but generally I'd say yes. For example, if state of an imaginable sender is described for a receiver like this: `( a + b * c ) / d` and the receiver reconstructs it on it's side, then the only correct order of replaying is (obviously): `add b; mul c; add a; div d`. It can be resolved, of course, but requires special treatment.

> the player should just be able to replay all traits and effects into a morale instance and get the same outcome.

It's tricky to implement without a general observation mechanism in place, otherwise it's an ugly special-case. Morale doesn't depend on traits and effects only. It requires worn items, body temps, wetness e.t.c. I really like the idea of the pattern, use it, but now I see, that it's too early to use it here, at first we should prepare soil.

codemime commented:
I've committed how it could be (IMHO pretty neat), however it won't work without proper invocations from the Observables. In order to make it work the `Creature`, `character` and `player` classes (the said "soil") should be modified to carefully track their state (they don't) and to support general observable interface. Suggestions are welcome. Meanwhile I'm going to write a bunch of unit tests for the `player_morale` class (I got the hint).

kevingranade commented on src/morale.cpp:
```diff
@@ -4,8 +4,16 @@
 #include "itype.h"
 #include "output.h"
+#include "bodypart.h"
+#include "player.h"
```
Looks like you can get rid of this include now.

kevingranade replied:
Or not, since it's calling player via the game object.

kevingranade commented on src/morale.cpp:
```diff
@@ -115,21 +203,22 @@ void morale_point::add( int new_bonus, int new_max_bonus, int new_duration, int
         bonus = new_max_bonus;
     }
+
+    age = 0; // Brand new. The assignment should stay below.
```
Comment is confusing, stay below what?

kevingranade commented on src/morale.h:
```diff
@@ -73,122 +81,120 @@ enum morale_type : int {
+        void display();
+
+    protected:
+        friend class player;
```
I'd rather these just be public, this is going to trip you up when you want to unit test it.

codemime replied:
Usually, I'd rather use inheritance to make protected methods accessible than modify the initial class to make the tests happy. Either way, this coupling with the player is supposed to be temporary.

kevingranade commented on src/morale.h:
```diff
@@ -4,8 +4,16 @@
 struct itype;
+struct morale_mult;
 
 enum morale_type : int {
```
It looks like morale.h is included in a lot of places, probably only to provide this enum (I noticed because I( changed something in morale.h and rebuilt, and it's rebuilding a significant fraction of the game).  Probably worthwhile to split it out into its own header file.

codemime replied:
It's a very good suggestion indeed.

codemime commented:
Added "events" for:
* gaining/losing mutations
* wearing/taking off items
* changing effect intensities
 
Transferred from `player` to `player_morale` (_further transfers are planned in forthcoming PRs_):
* body temperature morale (except comfy bonus)
* stylish morale
  
Added unit-tests for `player_morale`

Morale types are extracted into a separate header, but it didn't help much, because morale.h is included in player.h which causes recompilation.

Behavior remains the same (except mentioned fix).
